src/utils/post_utils.js

```diff
@@ -175,3 +183,82 @@ export function comparePosts(a, b) {
 
     return 0;
 }
+
+export function combineSystemPosts(postsIds = [], posts = {}, channelId) {
+    if (postsIds.length === 0) {
+        return {postsForChannel: postsIds, nextPosts: posts};
+    }
+
+    const postsForChannel = [];
+    const nextPosts = {...posts};
+
+    let userActivitySystemPosts = [];
+    let systemPostIds = [];
+    let createAt;
+    let combinedPostId;
+
+    postsIds.forEach((p, i) => {
+        const channelPost = posts[p];
+        if (
+            channelPost.delete_at === 0 &&
+            (isUserActivityPost(channelPost.type) || channelPost.type === Posts.POST_TYPES.COMBINED_USER_ACTIVITY)
+        ) {
+            if (!createAt || createAt > channelPost.create_at) {
+                createAt = channelPost.create_at;
+            }
+
+            if (isUserActivityPost(channelPost.type)) {
+                userActivitySystemPosts = [...userActivitySystemPosts, channelPost];
+                systemPostIds.push(channelPost.id);
+
+                if (nextPosts[channelPost.id]) {
+                    nextPosts[channelPost.id] = {...channelPost, state: Posts.POST_DELETED, delete_at: 1};
+                }
+            } else if (channelPost.type === Posts.POST_TYPES.COMBINED_USER_ACTIVITY) {
+                userActivitySystemPosts = [...userActivitySystemPosts, ...channelPost.user_activity_posts];
+                systemPostIds = [...systemPostIds, ...channelPost.system_post_ids];
+
+                combinedPostId = channelPost.id;
+            }
+        }
+
+        if (
+            (channelPost.type === '' && userActivitySystemPosts.length > 0) ||
+            userActivitySystemPosts.length === MAX_COMBINED_SYSTEM_POSTS ||
+            (userActivitySystemPosts.length > 0 && i === postsIds.length - 1)
+        ) {
+            const combinedPost = {
+                id: combinedPostId || generateId(),
+                root_id: '',
+                channel_id: channelId,
+                type: Posts.POST_TYPES.COMBINED_USER_ACTIVITY,
+                message: '',
+                create_at: createAt,
+                delete_at: 0,
+                user_activity_posts: userActivitySystemPosts,
+                system_post_ids: systemPostIds,
+                state: '',
+            };
+
+            nextPosts[combinedPost.id] = combinedPost;
+            postsForChannel.push(combinedPost.id);
+
+            userActivitySystemPosts = [];
+            systemPostIds = [];
+            createAt = null;
+            combinedPostId = null;
+
+            if (channelPost.type === '') {
+                postsForChannel.push(channelPost.id);
+            }
+        } else if (channelPost.type === '') {
+            postsForChannel.push(channelPost.id);
+        }
+    });
```

Comment:
after adding the new combined post to the `postsForChannel` you should remove those that were marked as `DELETED`? or do you see a better way to accomplish that?

Reply:
I'm just marking those system posts as deleted so that for the next round of received post/s those are known as received already and no need to merge with existing combinedPost. I think it's straightforward rather than checking each time whether the combinedPost has already the system post or not.